### PR TITLE
ingester: use more efficient spanlogger.DebugLog call

### PR DESCRIPTION
**Pull request by bboreham**
#### What this PR does

Reduce overhead of debug logging in ingester push requests, via a change to dskit SpanLogger 
https://github.com/grafana/dskit/pull/367.

I modified the benchmark to send 10 series instead of 10,000 - this is much more like what we get with typical Prometheus senders in a large deployment (500 samples sent x3 replication / 150 ingesters).

```
goos: linux
goarch: amd64
pkg: github.com/grafana/mimir/pkg/ingester
cpu: Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz
               │  before.txt  │              after.txt              │
               │    sec/op    │    sec/op     vs base               │
IngesterPush-4   26.24µ ± 21%   21.42µ ± 10%  -18.37% (p=0.002 n=6)

               │  before.txt  │              after.txt              │
               │     B/op     │     B/op      vs base               │
IngesterPush-4   3.021Ki ± 1%   1.119Ki ± 1%  -62.96% (p=0.002 n=6)

               │ before.txt │             after.txt             │
               │ allocs/op  │ allocs/op   vs base               │
IngesterPush-4   51.00 ± 0%   26.00 ± 0%  -49.02% (p=0.002 n=6)
```

#### Checklist

- NA Tests updated
- NA Documentation added
- NA `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/log/log.go [hunk 1]
```diff
@@ -47,13 +47,30 @@ func InitLogger(logFormat string, logLevel dslog.Level, buffered bool, rateLimit
 		logger = log.With(logger, "ts", log.DefaultTimestampUTC, "caller", log.Caller(5))
 	}
 	// Must put the level filter last for efficiency.
-	logger = level.NewFilter(logger, logLevel.Option)
+	Logger = newFilter(logger, logLevel)
```
Comment (jhalterman) [inline pkg/util/log/log.go:50]: Is this meant to be an assignment to `logger`?

Comment (bboreham) [reply pkg/util/log/log.go:50]: Yes, good catch.  Got lost in merge conflicts.

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, provided the dskit PR is merged and you remove the replace directive in go.mod

File: pkg/util/log/log.go [hunk 2]
```diff
@@ -47,13 +47,30 @@ func InitLogger(logFormat string, logLevel dslog.Level, buffered bool, rateLimit
...
-	logger = level.NewFilter(logger, logLevel.Option)
+	logger = newFilter(logger, logLevel)
 
 	// Set global logger.
 	Logger = logger
 	return logger
 }
 
+// Pass through Logger and implement the DebugEnabled interface that spanlogger looks for.
+type levelFilter struct {
```
Comment (pracucci) [inline pkg/util/log/log.go:58]: This looks worth to be moved to dskit too, so that other projects can use it too.

File: go.mod [hunk 3]
```diff
@@ -266,3 +266,6 @@ replace github.com/munnerz/goautoneg => github.com/charleskorn/goautoneg v0.0.0-
 
 // Replace opentracing-contrib/go-stdlib with a fork until https://github.com/opentracing-contrib/go-stdlib/pull/68 is merged.
 replace github.com/opentracing-contrib/go-stdlib => github.com/grafana/opentracing-contrib-go-stdlib v0.0.0-20230509071955-f410e79da956
+
+// Temporary until https://github.com/grafana/dskit/pull/367 is merged.
+replace github.com/grafana/dskit => github.com/grafana/dskit v0.0.0-20230822104759-dbdd569cac93
```
Comment (pracucci) [inline go.mod:271]: Make sure to merge dskit PR and remove this before merging Mimir PR, please.

File: pkg/ingester/ingester_test.go [hunk 4]
```diff
@@ -1409,8 +1409,8 @@ func BenchmarkIngesterPush(b *testing.B) {
 	require.NoError(b, err)
 
 	const (
-		series  = 10000
-		samples = 10
+		series  = 10
```
Comment (pracucci) [inline pkg/ingester/ingester_test.go:1412]: Last time I checked, the average we see at Grafana Labs was around 300. I would increase it.

**Review by pracucci — APPROVED** [review]
Nice optimization. I left a comment in dskit PR wondering if we can avoid adding `DebugLog()` log at all, and just infer the level from `Log()`.

File: pkg/ingester/ingester_test.go (hunk 4 above)
Comment (bboreham) [reply pkg/ingester/ingester_test.go:1412]: Are you thinking of distributors?  I did 
```
sum (rate(cortex_ingester_ingested_samples_total[5m])) / 
sum (rate(cortex_request_duration_seconds_count{route="/cortex.Ingester/Push",status_code="success"}[5m]))
```

and got 8.4.

**Pull request merged**